<!DOCTYPE html PUBLIC '-//W3C//DTD HTML 4.01//EN' 'http://www.w3.org/TR/html4/strict.dtd'><html><head> <meta http-equiv='content-type' content='text/html; charset=ISO-8859-1'> <link href='favicon.ico' rel='shortcut icon'><title>Getting Started</title></head> <body style='background-color: white; color: rgb(0, 0, 0);' alink='#000099' link='#000099' vlink='#990099'> <div id='logo'><div align='right'><div style='text-align: left;'><table style='text-align: left; width: 1073px; height: 805px;' border='0' cellpadding='0' cellspacing='0'> <tbody> <tr> <td style='width: 951px;'><div style='text-align: center;'><div style='text-align: left;'><big style='color: black;'><img style='width: 200px; height: 109px;' src='logo.png' alt=''><span style='font-weight: bold;'> </span></big><span style='font-weight: bold;'></span><br><span style='font-weight: bold;'><small style='color: rgb(102, 102, 102);'><big style='color: black;'><big><big><span style='font-family: Courier New,Courier,monospace;'>http://bioinfo.uqam.ca/armadillo</span></big></big></big></small></span><span style='font-weight: bold;'><small style='color: rgb(102, 102, 102);'><big style='color: black;'><big><big><span style='font-family: Courier New,Courier,monospace;'><br>armadillo.workflow@gmail.com</span></big></big></big></small></span><br></div><span style='font-weight: bold;'></span></div><span class='Apple-style-span' style='border-collapse: separate; color: rgb(0, 0, 0); font-family: 'Times New Roman'; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; font-size: medium; background-color: white;'><span id='internal-source-marker_0.49239397584460676' style='font-size: 11pt; font-family: Arial; color: rgb(0, 0, 0); font-weight: normal; font-style: normal; text-decoration: none; vertical-align: baseline;'><span style='font-weight: bold;'><br>Armadillo, a simple approach of automating comparative genomics tasks using workflows.</span><br>Etienne Lord, Mickael Leclercq, Alix Boc, Abdoulaye Banir√© Diallo, Vladimir Makarenkov<br>Department of Computer Science, Universite du Quebec a Montreal, Montreal, PO Box 8888<br>Downtown Station, Montreal, QC, H3C 3P8</span></span><span style='font-weight: bold;'><small style='color: rgb(102, 102, 102);'><big style='color: black;'><br><br></big></small></span><div style='text-align: justify;'><span class='Apple-style-span' style='border-collapse: separate; color: rgb(0, 0, 0); font-family: 'Times New Roman'; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; font-size: medium; background-color: white;'><div style='margin: 0px;'><span id='internal-source-marker_0.49239397584460676' style='font-size: 11pt; font-family: Arial; color: rgb(0, 0, 0); font-weight: normal; font-style: normal; text-decoration: none; vertical-align: baseline;'>The Armadillo workflow platform was first developed as an educational tool to help students understand the basic phylogenetic inference process using the PHYLIP package. However, it can now help life science researcher&#8217;s rapidly develop data-flux prototype and easily automate the analysis of multiple species phylogeny in an automated fashion, on a workstation. The current version (version 1.0) includes custom interface for software&#8217;s used in phylogenetic analysis such as search for orthologous genes, multiple sequence alignment (MSA), evolution model determination and inference of phylogeny using maximum likelihood or maximum parsimony. The workflows and their associated data are saved to a unique local database file, allowing the easy sharing of the developed workflow, annotation, analysis, and pipeline execution results to other re-searchers. Since the file is by itself a database, it can also be interrogated by using the database standard structured-query language (SQL). Additionally, this open-source software allows the scientific community to develop their own modules or to integrate other software&#8217;s into this application for their rapidstudy prototyping.</span></div></span></div><span style='font-weight: bold;'><small style='color: rgb(102, 102, 102);'><big style='color: black;'><br></big></small></span><div style='text-align: center;'><div style='text-align: left;'><span style='font-weight: bold;'><small style='color: rgb(102, 102, 102);'><big style='color: black;'><img style='width: 913px; height: 368px;' alt='' src='image.png'></big></small></span><br></div><div style='text-align: left;'><span class='Apple-style-span' style='border-collapse: separate; color: rgb(0, 0, 0); font-family: 'Times New Roman'; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; font-size: medium; background-color: white;'><span id='internal-source-marker_0.49239397584460676' style='font-size: 11pt; font-family: Arial; color: rgb(0, 0, 0); font-weight: normal; font-style: normal; text-decoration: none; vertical-align: baseline;'>Sample workflow created with the <span style='font-style: italic;'>Armadillo</span> workflow platform. Multiple protein sequences are first aligned with Muscle before a phylogenetic tree is inferred with PhyML or with ProtTest (PhyML) with different evolutionnary models. Finally, a tree distance is inferred with the Robinson and Fould methodology.</span></span><br><br><span class='Apple-style-span' style='border-collapse: separate; color: rgb(0, 0, 0); font-family: 'Times New Roman'; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; font-size: medium; background-color: white;'><span id='internal-source-marker_0.49239397584460676' style='font-size: 11pt; font-family: Arial; color: rgb(0, 0, 0); font-weight: normal; font-style: normal; text-decoration: none; vertical-align: baseline;'></span></span></div><hr style='width: 100%; height: 2px;'><div style='text-align: left;'><span class='Apple-style-span' style='border-collapse: separate; color: rgb(0, 0, 0); font-family: 'Times New Roman'; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; font-size: medium; background-color: white;'><span id='internal-source-marker_0.49239397584460676' style='font-size: 11pt; font-family: Arial; color: rgb(0, 0, 0); font-weight: normal; font-style: normal; text-decoration: none; vertical-align: baseline;'></span></span><span class='Apple-style-span' style='border-collapse: separate; color: rgb(0, 0, 0); font-family: 'Times New Roman'; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; font-size: medium; background-color: white;'><span id='internal-source-marker_0.49239397584460676' style='font-size: 11pt; font-family: Arial; color: rgb(0, 0, 0); font-weight: normal; font-style: normal; text-decoration: none; vertical-align: baseline;'><br><span style='font-weight: bold;'>Tutorials</span><br><br>Tutorials can be found on the website: <span style='font-weight: bold; font-family: Courier New,Courier,monospace;'>http://bioinfo.uqam.ca/armadillo</span><span style='font-family: Courier New,Courier,monospace;'></span><br></span></span><span class='Apple-style-span' style='border-collapse: separate; color: rgb(0, 0, 0); font-family: 'Times New Roman'; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; font-size: medium; background-color: white;'><span id='internal-source-marker_0.49239397584460676' style='font-size: 11pt; font-family: Arial; color: rgb(0, 0, 0); font-weight: normal; font-style: normal; text-decoration: none; vertical-align: baseline;'></span></span></div><span style='font-weight: bold;'></span></div><span style='font-weight: bold;'><small style='color: rgb(102, 102, 102);'><big style='color: black;'><br></big></small></span><span style='font-weight: bold;'></span><hr style='width: 100%; height: 2px;'><span style='font-weight: bold;'></span></td></tr><tr><td style='vertical-align: top; width: 951px;'><span class='Apple-style-span' style='border-collapse: separate; color: rgb(0, 0, 0); font-family: 'Times New Roman'; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; font-size: medium; background-color: white;'><span id='internal-source-marker_0.49239397584460676' style='font-size: 11pt; font-family: Arial; color: rgb(0, 0, 0); font-weight: normal; font-style: normal; text-decoration: none; vertical-align: baseline;'>Armadillo make use of the following libraries (and their associated packages): </span></span><br style='font-family: Arial;'><br><table x:str='' style='border-collapse: collapse; width: 1046px; height: 124px;' border='1' cellpadding='0' cellspacing='0'><col style='width: 285pt;' width='380'> <tbody><tr><td style='text-align: center; font-style: italic; font-weight: bold; background-color: rgb(153, 153, 153);'>Libraries</td><td style='text-align: center; font-style: italic; font-weight: bold; background-color: rgb(153, 153, 153);'>Version</td><td style='text-align: center; font-style: italic; font-weight: bold; background-color: rgb(153, 153, 153);'>Links</td></tr><tr style='height: 23.25pt;' height='31'><td class='xl68' style='border-top: medium none; height: 23.25pt; text-align: center; font-weight: bold;'>Processing</td><td class='xl69' style='border-top: medium none; border-left: medium none; text-align: center; font-weight: bold;'>1.2.1</td><td class='xl70' style='border-top: medium none; border-left: medium none; text-align: center;'>http://processing.org/</td></tr><tr style='height: 23.25pt;' height='31'><td class='xl68' style='border-top: medium none; height: 23.25pt; text-align: center; font-weight: bold;' height='31'>BioJava</td><td class='xl69' style='border-top: medium none; border-left: medium none; text-align: center; font-weight: bold;'>1.7.1</td><td class='xl70' style='border-top: medium none; border-left: medium none; text-align: center;'>http://biojava.org/wiki/Main_Page</td></tr><tr style='height: 23.25pt;' height='31'><td class='xl68' style='border-top: medium none; height: 23.25pt; text-align: center; font-weight: bold;' height='31'>ReadSeq</td><td class='xl69' style='border-top: medium none; border-left: medium none; text-align: center; font-weight: bold;'>2.1.27</td><td class='xl70' style='border-top: medium none; border-left: medium none; text-align: center;'>http://iubio.bio.indiana.edu/soft/molbio/readseq/java/</td></tr><tr style='height: 23.25pt;' height='31'><td class='xl68' style='border-top: medium none; height: 23.25pt; text-align: center; font-weight: bold;' height='31'>XerialSQLite JDBC Driver</td><td class='xl69' style='border-top: medium none; border-left: medium none; text-align: center; font-weight: bold;'>3.6.20.1</td><td style='border-top: medium none; border-left: medium none; text-align: center;'>http://www.xerial.org/trac/Xerial/wiki/SQLiteJDBC</td></tr></tbody></table><br>
<small></small><br><hr style='width: 100%; height: 2px;'><span class='Apple-style-span' style='border-collapse: separate; color: rgb(0, 0, 0); font-family: 'Times New Roman'; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; font-size: medium; background-color: white;'><span id='internal-source-marker_0.49239397584460676' style='font-size: 11pt; font-family: Arial; color: rgb(0, 0, 0); font-weight: normal; font-style: normal; text-decoration: none; vertical-align: baseline;'>Armadillo includes the following executables in its distribution: </span></span><br><br><table x:str='' style='border-collapse: collapse; width: 1062px; height: 3165px;' border='1' cellpadding='0' cellspacing='0'><col span='7' style='width: 60pt;' width='80'> <tbody><tr><td style='height: 24px; background-color: rgb(153, 153, 153); font-weight: bold; text-align: center; font-style: italic;'>Software</td><td style='height: 24px; background-color: rgb(153, 153, 153); font-weight: bold; text-align: center; font-style: italic;'>Version</td><td style='height: 24px; background-color: rgb(153, 153, 153); font-weight: bold; text-align: center; font-style: italic;'>Type</td><td style='width: 407px; height: 24px; background-color: rgb(153, 153, 153); font-weight: bold; text-align: center; font-style: italic;'>References</td><td style='width: 301px; height: 24px; background-color: rgb(153, 153, 153); font-weight: bold; text-align: center; font-style: italic;' colspan='1' rowspan='1'>Website</td></tr><tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>ClustalW2</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.1</td>
<td style='width: 60pt; text-align: center;' width='80'>Alignment(s)</td>
<td style='width: 407px;'><small>Larkin MA, Blackshields G, Brown NP, Chenna R, McGettigan PA, McWilliam H, Valentin F, Wallace IM, Wilm A, Lopez R, Thompson JD, Gibson TJ, Higgins DG. 
(2007). Clustal W and Clustal X version 2.0. Bioinformatics, 23, 2947-2948.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.clustal.org/'>http://www.clustal.org/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>MAFFT</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.8.31</td>
<td style='width: 60pt; text-align: center;' width='80'>Alignment(s)</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href=''></a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Extract Sequences</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.1</td>
<td style='width: 60pt; text-align: center;' width='80'>Alignment(s)</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href=''></a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Probcons</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.12</td>
<td style='width: 60pt; text-align: center;' width='80'>Alignment(s)</td>
<td style='width: 407px;'><small>Do CB, Mahabhashyam MS, Brudno M, Batzoglou S (2005) ProbCons: Probabilistic consistency-based multiple sequence alignment. Genome Res. 15: 330-40.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://probcons.stanford.edu/'>http://probcons.stanford.edu/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>ClustalW</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.83</td>
<td style='width: 60pt; text-align: center;' width='80'>Alignment(s)</td>
<td style='width: 407px;'><small>Thompson JD, Higgins DG, Gibson TJ. 
(1994). CLUSTAL W: improving the sensitivity of progressive multiple sequence alignment through sequence weighting, position-specific gap penalties and weight matrix choice. Nucleic Acids Res., 22, 4673-4680.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.clustal.org/'>http://www.clustal.org/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>GBlocks</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>0.91b</td>
<td style='width: 60pt; text-align: center;' width='80'>Alignment(s)</td>
<td style='width: 407px;'><small>Talavera, G., and Castresana, J. (2007). Improvement of phylogenies after removing divergent and ambiguously aligned blocks from protein sequence alignments. Systematic Biology 56, 564-577. 

Castresana, J. (2000). Selection of conserved blocks from multiple alignments for their use in phylogenetic analysis. Molecular Biology and Evolution 17, 540-552.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://molevol.cmima.csic.es/castresana/Gblocks.html'>http://molevol.cmima.csic.es/castresana/Gblocks.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Kalign</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.03</td>
<td style='width: 60pt; text-align: center;' width='80'>Alignment(s)</td>
<td style='width: 407px;'><small>Timo Lassmann and Erik LL Sonnhammer (2005) 
Kalign - an accurate and fast multiple sequence alignment algorithm 
BMC Bioinformatics 2005, 6:298 doi:10.1186/1471-2105-6-298)</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.ebi.ac.uk/Tools/msa/kalign/'>http://www.ebi.ac.uk/Tools/msa/kalign/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Muscle</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.8.31</td>
<td style='width: 60pt; text-align: center;' width='80'>Alignment(s)</td>
<td style='width: 407px;'><small>Edgar, R.C. (2004) MUSCLE: multiple sequence alignment with high accuracy and high throughput.Nucleic Acids Res. 32(5):1792-1797.
doi:10.1093/nar/gkh340

Edgar, R.C. (2004) MUSCLE: a multiple sequence alignment method with reduced time and space complexity BMC Bioinformatics, (5) 113.
doi:10.1186/1471-2105-5-113</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.drive5.com/muscle/'>http://www.drive5.com/muscle/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Local Blast</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.2.25</td>
<td style='width: 60pt; text-align: center;' width='80'>Blast</td>
<td style='width: 407px;'><small>BLAST+: architecture and applications.
Camacho C, Coulouris G, Avagyan V, Ma N, Papadopoulos J, Bealer K, Madden TL.
BMC Bioinformatics. 2009 Dec 15;10:421</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download'>http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Blast (EBI)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.0</td>
<td style='width: 60pt; text-align: center;' width='80'>Blast</td>
<td style='width: 407px;'><small>NA</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.ncbi.nlm.nih.gov/blast/Doc/urlapi.html'>http://www.ncbi.nlm.nih.gov/blast/Doc/urlapi.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Create Local BlastDB</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.2.25</td>
<td style='width: 60pt; text-align: center;' width='80'>Blast</td>
<td style='width: 407px;'><small>BLAST+: architecture and applications.
Camacho C, Coulouris G, Avagyan V, Ma N, Papadopoulos J, Bealer K, Madden TL.
BMC Bioinformatics. 2009 Dec 15;10:421</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download'>http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Weighted K-Means</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.0</td>
<td style='width: 60pt; text-align: center;' width='80'>Clustering</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href=''></a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Convert SVG to PNG</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.7</td>
<td style='width: 60pt; text-align: center;' width='80'>Conversion</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://xmlgraphics.apache.org/batik/'>http://xmlgraphics.apache.org/batik/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>BWA</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>0.7.10</td>
<td style='width: 60pt; text-align: center;' width='80'>Genomic</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>EMBOSS chips</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: EMBOSS</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href=''></a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>EMBOSS einverted</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: EMBOSS</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href=''></a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>EMBOSS aaindexextract</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: EMBOSS</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href=''></a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>EMBOSS sizeseq</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: EMBOSS</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>EMBOSS diffseq</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: EMBOSS</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href=''></a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>EMBOSS banana</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: EMBOSS</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href=''></a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>EMBOSS est2genome</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: EMBOSS</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href=''></a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Bowtie</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.1.2</td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Index Genome</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>BWA Index</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Index Genome</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Bowtie2</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.2.5</td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Index Genome</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Bowtie2</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.2.5</td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Inspect Indexed Genome</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Bowtie</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.1.2</td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Inspect Indexed Genome</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>BWA</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Mapping</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>miRdup</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Mapping</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Bowties</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Mapping</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Bowtie</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.1.2</td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Mapping</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Bowtie2</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.2.5</td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Mapping</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>MAQ</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>0.7.1</td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Mapping</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://maq.sourceforge.net/'>http://maq.sourceforge.net/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>FastQC</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>0.11.3</td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: QC & Manip</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.bioinformatics.babraham.ac.uk/projects/fastqc/'>http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Cutadapt</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: QC and Manipulation</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>miRCheck</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: RNA</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Tapir</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: RNA</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>RepeatMasker</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: RepeatMasker</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.repeatmasker.org/faq.html'>http://www.repeatmasker.org/faq.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>RNAfold</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: Vienna</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>miRCheck fold inverted repeats</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: miRcheck</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>miRCheck retreive genomic regions</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: miRcheck</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>miRCheck einverted</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: miRcheck</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>miRCheck extract einverted 20mers</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: miRcheck</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>miRCheck patscan</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: miRcheck</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>miRCheck_evaluate_miRNA_candidates</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>NGS: miRcheck</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>BAli-phy</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.11</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Alignment</td>
<td style='width: 407px;'><small>Suchard MA and Redelings BD BAli-Phy: simultaneous Bayesian inference of alignment and phylogeny, Bioinformatics, 22:2047-2048, 2006.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.biomath.ucla.edu/msuchard/bali-phy/'>http://www.biomath.ucla.edu/msuchard/bali-phy/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Alignment information</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.11</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Alignment</td>
<td style='width: 407px;'><small>Suchard MA and Redelings BD BAli-Phy: simultaneous Bayesian inference of alignment and phylogeny, Bioinformatics, 22:2047-2048, 2006.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.bali-phy.org'>http://www.bali-phy.org</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>AncestorCC</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.0</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Ancestral Reconstruction</td>
<td style='width: 407px;'><small>Ancestors 1.0: a web server for ancestral sequence reconstruction.
Diallo AB, Makarenkov V, Blanchette M.
Bioinformatics. 2010 Jan 1;26(1):130-1. Epub 2009 Oct 22.

Exact and heuristic algorithms for the Indel Maximum Likelihood Problem.
Diallo AB, Makarenkov V, Blanchette M.
J Comput Biol. 2007 May;14(4):446-61.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://ancestors.bioinfo.uqam.ca/ancestorWeb/'>http://ancestors.bioinfo.uqam.ca/ancestorWeb/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>jModelTest (Nucleic Acid)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>0.1</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Evolutionary Model Testing</td>
<td style='width: 407px;'><small>Selection of models of DNA evolution with jModelTest.
Posada D.
Methods Mol Biol. 2009;537:93-112.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://darwin.uvigo.es/software/jmodeltest.html'>http://darwin.uvigo.es/software/jmodeltest.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>ProtTest (Amino Acid)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.4</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Evolutionary Model Testing</td>
<td style='width: 407px;'><small>Abascal F, Zardoya R, Posada, D. 2005. ProtTest: Selection of best-fit models of protein evolution. Bioinformatics: 21(9):2104-2105.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://darwin.uvigo.es/software/prottest.html'>http://darwin.uvigo.es/software/prottest.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>PIRN</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.0</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Horizontal Genes Tranfer</td>
<td style='width: 407px;'><small>Yufeng Wu, "Close lower and upper bounds for the minimum reticulate network of multiple phylogenetic trees", in Proceedings of ISMB 2010, Bioinformatics 2010 26(12):i140-i148, 2010</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.engr.uconn.edu/~ywu/PIRN.html'>http://www.engr.uconn.edu/~ywu/PIRN.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>SPR Identification Tool</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.0</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Horizontal Genes Tranfer</td>
<td style='width: 407px;'><small>SPRIT: Identifying horizontal gene transfer in rooted phylogenetic trees.
Hill T, Nordstr√∂m KJ, Thollesson M, S√§fstr√∂m TM, Vernersson AK, Fredriksson R, Schi√∂th HB.
BMC Evol Biol. 2010 Feb 13;10:42.
PMID: 20152048</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://code.google.com/p/phylogenetics/'>http://code.google.com/p/phylogenetics/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>HGT Detector 3.2</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.2</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Horizontal Genes Tranfer</td>
<td style='width: 407px;'><small>Inferring and validating horizontal gene transfer events using bipartition dissimilarity.
Boc A, Philippe H, Makarenkov V.
Syst Biol. 2010 Mar;59(2):195-211.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='trex.bioinfo.uqam.ca'>trex.bioinfo.uqam.ca</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>SPRDist</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.0.2</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Horizontal Genes Tranfer</td>
<td style='width: 407px;'><small>A practical method for exact computation of subtree prune and regraft distance.
Wu Y.
Bioinformatics. 2009 Jan 15;25(2):190-6. Epub 2008 Nov 19.
PMID: 19019848</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.engr.uconn.edu/~ywu/SPRDist.html'>http://www.engr.uconn.edu/~ywu/SPRDist.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>LatTrans</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2003</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Horizontal Genes Tranfer</td>
<td style='width: 407px;'><small>Louigi Addario-Berry, Michael T. Hallett, and Jens Lagergren, Towards Identifying Lateral Gene Transfer Events. Pacific Symposium on Biocomputing 2003: 279-290</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.math.mcgill.ca/louigi/'>http://www.math.mcgill.ca/louigi/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>EEEP</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.01</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Horizontal Genes Tranfer</td>
<td style='width: 407px;'><small>Beiko, R.G. and Hamilton, N. (2006). Phylogenetic identification of lateral genetic transfer events. BMC Evolutionary Biology 6:15.
Beiko, R.G., Harlow, T.J., and Ragan, M.A. (2005). Highways of gene sharing in prokaryotes. Proc. Natl. Acad. Sci. USA 102:14332-14337.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://acb.qfab.org/acb/eeep/'>http://acb.qfab.org/acb/eeep/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Hybrid Interleave</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.0</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Horizontal Genes Tranfer</td>
<td style='width: 407px;'><small>Publication(s)Quantifying hybridization in realistic time.
Collins J, Linz S, Semple C.
J Comput Biol. 2011 Oct;18(10):1305-18. Epub 2011 Jan 6.
PMID: 21210735</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://ab.inf.uni-tuebingen.de/people/linz/software/HybridInterleave.zip/view'>http://ab.inf.uni-tuebingen.de/people/linz/software/HybridInterleave.zip/view</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Ranger-dtl</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.0</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Horizontal Genes Tranfer</td>
<td style='width: 407px;'><small>Efficient Algorithms for the Reconciliation Problem with Gene Duplication, Horizontal Transfer, and Loss
Mukul S. Bansal, Eric J. Alm, Manolis Kellis. 
ISMB 2012; Bioinformatics 28: i283-i291, 2012.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://compbio.mit.edu/ranger-dtl/'>http://compbio.mit.edu/ranger-dtl/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>PhyloNet - RiataHGT</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.4</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Horizontal Genes Tranfer</td>
<td style='width: 407px;'><small>L. Nakhleh, D. Ruths, and L. S. Wang. 
RIATA-HGT: A Fast and Accurate Heuristic for Reconstruction Horizontal Gene
Proceedings of the 11th International Computing and Combinatorics Conference (COCOON 05). LNCS #3595 (L. Wang, editor), 84-93. 2005.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://bioinfo.cs.rice.edu/phylonet/'>http://bioinfo.cs.rice.edu/phylonet/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Seq-Gen</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.3.2</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Random Sequences or Trees</td>
<td style='width: 407px;'><small>Seq-Gen: an application for the Monte Carlo simulation of DNA sequence evolution along phylogenetic trees. Rambaut A, Grassly NC.
Comput Appl Biosci. 1997 Jun;13(3):235-8.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://tree.bio.ed.ac.uk/software/seqgen/'>http://tree.bio.ed.ac.uk/software/seqgen/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Random Tree</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Random Sequences or Trees</td>
<td style='width: 407px;'><small>(N. Bray and L. Pachter, MAVID: Constrained ancestral alignment of multiple sequences,, Genome Research, 14:693-699 (2004))</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://bio.math.berkeley.edu/mavid/'>http://bio.math.berkeley.edu/mavid/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Random Tree Generator</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Random Sequences or Trees</td>
<td style='width: 407px;'><small>Kuhner, M., and J. Felsenstein. 1994. A simulation comparison of phylogeny algorithms under equal and unequal evolutionary rates. Mol. Biol. Evol. 11:459-468.
Guindon, S., and O. Gascuel. 2002. Efficient biased estimation of evolutionary distances when substitution rates vary across sites. Mol. Biol. Evol. 19:534-543.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://trex.labunix.uqam.ca/index.php?action=randomtreegenerator&project=trex'>http://trex.labunix.uqam.ca/index.php?action=randomtreegenerator&project=trex</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>PaML (baseml)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>4.4</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Selective Pressure</td>
<td style='width: 407px;'><small>Yang Z (2007) PAML 4: phylogenetic analysis by maximum likelihood. Mol Biol Evol. 24: 1586-1591.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://abacus.gene.ucl.ac.uk/software/paml.html'>http://abacus.gene.ucl.ac.uk/software/paml.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Q function</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2011</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Selective Pressure</td>
<td style='width: 407px;'><small>A whole genome study and identification of specific carcinogenic regions of the human papilloma viruses.
Diallo AB, Badescu D, Blanchette M, Makarenkov V.
J Comput Biol. 2009 Oct;16(10):1461-73.
PMID: 19754274</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href=''></a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>PaML (codeml)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>4.4</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Selective Pressure</td>
<td style='width: 407px;'><small>Yang Z (2007) PAML 4: phylogenetic analysis by maximum likelihood. Mol Biol Evol. 24: 1586-1591.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://abacus.gene.ucl.ac.uk/software/paml.html'>http://abacus.gene.ucl.ac.uk/software/paml.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>PaML (yn00)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>4.4</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Selective Pressure</td>
<td style='width: 407px;'><small>(A) Nei-Gojobori (1986) method
Nei M, Gojobori T (1986) Simple methods for estimating the numbers of synonymous and nonsynonymous nucleotide substitutions. Mol. Biol. Evol. 3:418-426
(B) Yang & Nielsen (2000) method
Yang Z, Nielsen R (2000) Estimating synonymous and nonsynonymous substitution rates under realistic evolutionary models. Mol. Biol. Evol. 17:32-43
(C) LWL85, LPB93 & LWLm methods
Li W.-H., C.-I. Wu, Luo (1985) A new method for estimating synonymous and nonsynonymous rates of nucleotide substitutions considering the relative likelihood of nucleotide and codon changes. Mol. Biol. Evol. 2: 150-174.
Li W-H (1993) Unbiased estimation of the rates of synonymous and nonsynonymous substitution. J. Mol. Evol. 36:96-99
Pamilo P, Bianchi NO (1993) Evolution of the Zfx and Zfy genes - rates and interdependence between the genes. Mol. Biol. Evol. 10:271-281
Yang Z (2006) Computational Molecular Evolution. Oxford University Press, Oxford, England</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://abacus.gene.ucl.ac.uk/software/paml.html'>http://abacus.gene.ucl.ac.uk/software/paml.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>BIONJ</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.0</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small>"BIONJ: an improved version of the NJ algorithm based on a simple model of sequence data."
Gascuel O.
Molecular Biology and Evolution. 1997 14:685-695</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.atgc-montpellier.fr/bionj/'>http://www.atgc-montpellier.fr/bionj/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>FastTree2</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.1.7</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small>Price, M.N., Dehal, P.S., and Arkin, A.P. (2009) FastTree: Computing Large Minimum-Evolution Trees with Profiles instead of a Distance Matrix. Molecular Biology and Evolution 26:1641-1650, doi:10.1093/molbev/msp077.

Price, M.N., Dehal, P.S., and Arkin, A.P. (2010) FastTree 2 -- Approximately Maximum-Likelihood Trees for Large Alignments. PLoS ONE, 5(3):e9490. doi:10.1371/journal.pone.0009490.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://meta.microbesonline.org/fasttree/'>http://meta.microbesonline.org/fasttree/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Archaeopteryx - Viewer</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'> 0.957</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small>Han M.V. and Zmasek C.M. (2009). phyloXML: XML for evolutionary biology and comparative genomics. BMC Bioinformatics, 10:356.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.phylosoft.org/archaeopteryx/'>http://www.phylosoft.org/archaeopteryx/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Garli</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.0</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small>Zwickl, D. J., 2006. Genetic algorithm approaches for the phylogenetic analysis of large biological sequence datasets under the maximum likelihood criterion. Ph.D. dissertation,The University of Texas at Austin.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='https://www.nescent.org/wg_garli/Main_Page'>https://www.nescent.org/wg_garli/Main_Page</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>RootTree (MidPoint)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small>N. Bray and L. Pachter, MAVID: Constrained ancestral alignment of multiple sequences,, Genome Research, 14:693-699 (2004).</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://bio.math.berkeley.edu/mavid/'>http://bio.math.berkeley.edu/mavid/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>MrBayes</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.2.1</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small>MrBayes 3: Bayesian phylogenetic inference under mixed models.
Ronquist F, Huelsenbeck JP.
Bioinformatics. 2003 Aug 12;19(12):1572-4.

Altekar, G., S. Dwarkadas, J. P. Huelsenbeck, and F. Ronquist. 2004. Parallel Metropolis-coupled Markov chain Monte Carlo for Bayesian phylogenetic inference. Bioinformatics 20:407-415.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://molecularevolution.org/software/phylogenetics/mrbayes'>http://molecularevolution.org/software/phylogenetics/mrbayes</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>NINJA</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.2.1</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small>Wheeler, T.J. 2009. Large-scale neighbor-joining with NINJA. In S.L. Salzberg and T. Warnow (Eds.), Proceedings of the 9th Workshop on Algorithms in Bioinformatics. WABI 2009, pp. 375-389. Springer, Berlin.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://nimbletwist.com/software/ninja/'>http://nimbletwist.com/software/ninja/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Scriptree - Viewer</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>17</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small>ScripTree: scripting phylogenetic graphics.
Chevenet F, Croce O, Hebrard M, Christen R, Berry V.
Bioinformatics. 2010 Apr 15;26(8):1125-6.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://lamarck.lirmm.fr/scriptree/'>http://lamarck.lirmm.fr/scriptree/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Neighbor Joining</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>2.0</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href=''></a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>fastDNAml</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.2.2</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small>fastDNAmL: a tool for construction of phylogenetic trees of DNA sequences using maximum likelihood.
Olsen GJ, Matsuda H, Hagstrom R, Overbeek R.
Comput Appl Biosci. 1994 Feb;10(1):41-8.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://iubio.bio.indiana.edu/soft/molbio/evolve/fastdnaml/fastDNAml.html'>http://iubio.bio.indiana.edu/soft/molbio/evolve/fastdnaml/fastDNAml.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>PhyML</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.0.1</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small>"A simple, fast, and accurate algorithm to estimate large phylogenies by maximum likelihood."  Guindon S, Gascuel O.  Systematic Biology. 2003 52(5):696-704.  
This is a detailed description of the algorithm.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://code.google.com/p/phyml/'>http://code.google.com/p/phyml/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>RAxML</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.0.1</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree</td>
<td style='width: 407px;'><small>Alexandros Stamatakis : √íRAxML-VI-HPC: Maximum Likelihood-based Phylogenetic Analyses with Thousands of Taxa and Mixed Models√ì, Bioinformatics 22(21):2688√ê2690, 2006</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://www.exelixis-lab.org'>http://www.exelixis-lab.org</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>TreeDist (Phylip)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.69</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Distance</td>
<td style='width: 407px;'><small>Felsenstein, J. 1989. PHYLIP - Phylogeny Inference Package (Version 3.2). Cladistics 5: 164-166.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://evolution.genetics.washington.edu/phylip/'>http://evolution.genetics.washington.edu/phylip/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Robinson&Fould</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Distance</td>
<td style='width: 407px;'><small>Robinson, D.R. et L.R. Foulds. (1981). Comparison of phylogenetic trees. Math Biosci., volume 53, pages 131-147.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='trex.bioinfo.uqam.ca'>trex.bioinfo.uqam.ca</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>RETREE (Phylip)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.69</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Phylip</td>
<td style='width: 407px;'><small>PHYLIP -- Phylogeny Inference Package (Version 3.2). 
Felsenstein, J. 1989. Cladistics 5: 164-166</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://evolution.genetics.washington.edu/phylip.html'>http://evolution.genetics.washington.edu/phylip.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>PRODIST (Phylip)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.69</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Phylip</td>
<td style='width: 407px;'><small>Felsenstein, J. 1989. PHYLIP - Phylogeny Inference Package (Version 3.2). Cladistics 5: 164-166.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://evolution.genetics.washington.edu/phylip.html'>http://evolution.genetics.washington.edu/phylip.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>DNADIST (Phylip)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.69</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Phylip</td>
<td style='width: 407px;'><small>Felsenstein, J. 1989. PHYLIP - Phylogeny Inference Package (Version 3.2). Cladistics 5: 164-166.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://evolution.genetics.washington.edu/phylip.html'>http://evolution.genetics.washington.edu/phylip.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>CONSENSE (Phylip)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.69</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Phylip</td>
<td style='width: 407px;'><small>Felsenstein, J. 1989. PHYLIP - Phylogeny Inference Package (Version 3.2). Cladistics 5: 164-166</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://evolution.genetics.washington.edu/phylip.html'>http://evolution.genetics.washington.edu/phylip.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>PROTML (Phylip)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.69</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Phylip</td>
<td style='width: 407px;'><small>Felsenstein, J. 1989. PHYLIP - Phylogeny Inference Package (Version 3.2). Cladistics 5: 164-166.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://evolution.genetics.washington.edu/phylip.html'>http://evolution.genetics.washington.edu/phylip.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>SEQBOOT (Phylip)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.69</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Phylip</td>
<td style='width: 407px;'><small>Felsenstein, J. 1989. PHYLIP - Phylogeny Inference Package (Version 3.2). Cladistics 5: 164-166.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://evolution.genetics.washington.edu/phylip/'>http://evolution.genetics.washington.edu/phylip/</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>DNAML-Erate</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>1.0</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Phylip</td>
<td style='width: 407px;'><small>Probabilistic Phylogenetic Inference with Insertions and Deletions.
E. Rivas, S. R. Eddy.
PLoS Comput. Biol., 4:e1000172, 2008.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://selab.janelia.org/software.html'>http://selab.janelia.org/software.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>DNAML (Phylip)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.69</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Phylip</td>
<td style='width: 407px;'><small>Felsenstein, J. 1989. PHYLIP - Phylogeny Inference Package (Version 3.2). Cladistics 5: 164-166.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://evolution.genetics.washington.edu/phylip.html'>http://evolution.genetics.washington.edu/phylip.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>PROPARS (Phylip)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.69</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Phylip</td>
<td style='width: 407px;'><small>Felsenstein, J. 1989. PHYLIP - Phylogeny Inference Package (Version 3.2). Cladistics 5: 164-166.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://evolution.genetics.washington.edu/phylip.html'>http://evolution.genetics.washington.edu/phylip.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>NEIGHBOR (Phylip)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.69</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Phylip</td>
<td style='width: 407px;'><small>Felsenstein, J. 1989. PHYLIP - Phylogeny Inference Package (Version 3.2). Cladistics 5: 164-166.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://evolution.genetics.washington.edu/phylip.html'>http://evolution.genetics.washington.edu/phylip.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>DNAPARS (Phylip)</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'>3.69</td>
<td style='width: 60pt; text-align: center;' width='80'>Phy: Tree - Phylip</td>
<td style='width: 407px;'><small>Felsenstein, J. 1989. PHYLIP - Phylogeny Inference Package (Version 3.2). Cladistics 5: 164-166.</small></td>
<td rowspan='1' colspan='1' style='width: 301px;'><a href='http://evolution.genetics.washington.edu/phylip.html'>http://evolution.genetics.washington.edu/phylip.html</a></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Filter Sequences</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>Unknown</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
<tr>
<td style='height: 12.75pt; width: 60pt; text-align: center; font-weight: bold;' height='17' width='80'>Compare Distance</td>
<td style='width: 60pt; text-align: center; font-weight: bold;' width='80'></td>
<td style='width: 60pt; text-align: center;' width='80'>Unknown</td>
<td style='width: 407px;'><small></small></td>
<td rowspan='1' colspan='1' style='width: 301px;'></td></tr>
</tbody></table><br><span style='font-weight: bold;'><small style='color: rgb(102, 102, 102);'><big style='font-style: italic;'><br> </big></small></span> <hr style='width: 100%; height: 2px;'><span class='Apple-style-span' style='border-collapse: separate; color: rgb(0, 0, 0); font-family: 'Times New Roman'; font-style: normal; font-variant: normal; letter-spacing: normal; line-height: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; font-size: medium; background-color: white; font-weight: bold;'><span id='internal-source-marker_0.49239397584460676' style='font-size: 11pt; font-family: Arial; color: rgb(0, 0, 0); font-weight: normal; font-style: normal; text-decoration: none; vertical-align: baseline;'>Supported by grants from:</span></span><br> <span style='font-weight: bold;'></span><span style='font-weight: bold;'><small style='color: rgb(102, 102, 102);'><img style='border: 0px solid ; width: 150px; height: 49px;' alt='' src='uqam.png'></small></span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <img style='width: 200px; height: 89px;' alt='' src='crsng_logo.png'><span style='font-weight: bold;'><small style='color: rgb(102, 102, 102);'><br> </small></span><span style='font-weight: bold;'><small style='color: rgb(102, 102, 102);'><big><big><span style='color: black;'></span></big></big></small></span><hr style='width: 100%; height: 2px;'><span class='Apple-style-span' style='border-collapse: separate; color: rgb(0, 0, 0); font-family: 'Times New Roman'; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; font-size: medium; background-color: white;'><span id='internal-source-marker_0.49239397584460676' style='font-size: 11pt; font-family: Arial; color: rgb(0, 0, 0); font-weight: normal; font-style: normal; text-decoration: none; vertical-align: baseline;'>Licence informations: <span style='font-style: italic;'></span></span></span><div style='background-color: white; color: rgb(102, 102, 102);'><span style='font-weight: bold;'><small>This program is free software: you can redistribute it and/or modify</small></span><br> <span style='font-weight: bold;'><small>it under the terms of the GNU General Public License as published by</small></span><br> <span style='font-weight: bold;'><small>the Free Software Foundation, either version 3 of the License, or</small></span><br> <span style='font-weight: bold;'><small>(at your option) any later version.</small></span><br> <br> <span style='font-weight: bold;'><small>This program is distributed in the hope that it will be useful,</small></span><br> <span style='font-weight: bold;'><small>but WITHOUT ANY WARRANTY; without even the implied warranty of</small></span><br> <span style='font-weight: bold;'><small>MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.&nbsp; <br><br>See the </small></span><span style='font-weight: bold;'><small>GNU General Public License for more details.</small></span><br> <br><span style='font-weight: bold;'><small>You should have received a copy of the GNU General Public License</small></span><br> <span style='font-weight: bold;'><small>along with this program.&nbsp; If not, see&lt;<a href='http://www.gnu.org/licenses/'>http://www.gnu.org/licenses/</a>&gt;.</small></span></div></td></tr></tbody></table><span style='font-style: italic; font-weight: bold;'></span></div></div></div><span style='font-weight: bold;'><br></span></body></html> 
